pkg/controller/cni/cilium_reconciler.go

```diff
@@ -0,0 +1,245 @@
+	key := client.ObjectKey{
+		Namespace: capiCluster.Namespace,
+		Name:      capiCluster.Name,
+	}
+	remoteClient, err := cr.tracker.GetClient(ctx, key)
+	if err != nil {
+		return controller.Result{}, err
+	}
+
+	cr.Log.Info("Applying CNI")
+	ciliumDS := &v1.DaemonSet{}
+	ciliumDSName := types.NamespacedName{Namespace: "kube-system", Name: cilium.DaemonSetName}
+	err = remoteClient.Get(ctx, ciliumDSName, ciliumDS)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			cr.Log.Info("Deploying Cilium DS")
+			helm := executables.NewHelm(executables.NewExecutable("helm"), executables.WithInsecure())
+
+			ci := cilium.NewCilium(nil, helm)
+
+			ciliumSpec, err := ci.GenerateManifest(ctx, specWithBundles, []string{constants.CapvSystemNamespace})
+			if err != nil {
+				return controller.Result{}, err
+			}
+			if err := serverside.ReconcileYaml(ctx, remoteClient, ciliumSpec); err != nil {
+				return controller.Result{}, err
+			}
+			return controller.Result{}, err
+		}
+
+		return controller.Result{}, err
+	}
+
+	// upgrade cilium
+	cr.Log.Info("Upgrading Cilium")
+	needsUpgrade, err := ciliumNeedsUpgrade(ctx, cr.Log, remoteClient, specWithBundles)
+	if err != nil {
+		return controller.Result{}, err
+	}
+
+	if !needsUpgrade {
+		cr.Log.Info("Skipping Cilium")
+		return controller.Result{}, nil
+	}
+
+	helm := executables.NewHelm(executables.NewExecutable("helm"))
+	templater := cilium.NewTemplater(helm)
+	preflight, err := templater.GenerateUpgradePreflightManifest(ctx, specWithBundles)
+	if err != nil {
+		return controller.Result{}, err
+	}
+
+	cr.Log.Info("Installing Cilium upgrade preflight manifest")
+	if err := serverside.ReconcileYaml(ctx, remoteClient, preflight); err != nil {
+		return controller.Result{}, err
+	}
+
+	ciliumDS = &v1.DaemonSet{}
+	ciliumDSName = types.NamespacedName{Namespace: "kube-system", Name: "cilium"}
+	if err := remoteClient.Get(ctx, ciliumDSName, ciliumDS); err != nil {
+		cr.Log.Info("Cilium DS not found")
+		return controller.Result{}, err
+	}
+
+	if err := cilium.CheckDaemonSetReady(ciliumDS); err != nil {
+		cr.Log.Info("Cilium DS not ready")
+		return controller.Result{Result: &ctrl.Result{
+			RequeueAfter: defaultRequeueTime,
+		}}, nil
+	}
+
+	preFlightCiliumDS := &v1.DaemonSet{}
+	preFlightCiliumDSName := types.NamespacedName{Namespace: "kube-system", Name: cilium.PreflightDaemonSetName}
+	if err := remoteClient.Get(ctx, preFlightCiliumDSName, preFlightCiliumDS); err != nil {
+		cr.Log.Info("Preflight Cilium DS not found.")
+		return controller.Result{Result: &ctrl.Result{
+			RequeueAfter: defaultRequeueTime,
+		}}, err
+	}
+
+	if err := cilium.CheckPreflightDaemonSetReady(ciliumDS, preFlightCiliumDS); err != nil {
+		cr.Log.Info("Preflight DS not ready ")
+		return controller.Result{Result: &ctrl.Result{
+			RequeueAfter: defaultRequeueTime,
+		}}, err
+	}
+
+	cr.Log.Info("Deleting Preflight Cilium objects")
+	if err := serverside.DeleteYaml(ctx, remoteClient, preflight); err != nil {
```

Comment:
I think it might be safer to delete this after the upgrade manifest has been applied
that way we guarantee that the container images are not deleted

Reply:
done